File: C2/PDOO/Practicas/DeepSpaceJava/src/deepspace/Hangar.java


package deepspace;

import java.util.ArrayList;
/**
 *
 * @author Pedro Pablo Ruiz Huertas y Juan Antonio Villegas Recio
 */
public class Hangar implements Copyable <Hangar> {
    private int maxElements;
    private ArrayList<ShieldBooster> shieldBoosters=new ArrayList<>();
    private ArrayList<Weapon> weapons=new ArrayList<>();
    
    
    public Hangar(int capacity){
        maxElements=capacity;
    }
    
    public Hangar(Hangar h){
        weapons=h.weapons;
        shieldBoosters=h.shieldBoosters;
        maxElements=h.maxElements;
    }
    
    public HangarToUI getUIversion(){
        return new HangarToUI(this);
    }
    
    private boolean spaceAvailable(){
        
        return weapons.size()+shieldBoosters.size() < maxElements;
    }   
    
    public boolean addWeapon(Weapon w){
        if (this.spaceAvailable())
            return weapons.add(w);
        else return false;
    }
            
    public boolean addShieldBooster(ShieldBooster w){
        if (this.spaceAvailable())
            return shieldBoosters.add(w);
        else return false;
    }
    
    public int getMaxElements(){
        return maxElements;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return shieldBoosters;
    }
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    }
    
    public Weapon removeWeapon(int w){
        if (w>=weapons.size())
            return null;
        else{
            return weapons.remove(w);
        }
    }
    

    public ShieldBooster removeShieldBooster(int s){
        if (s>=shieldBoosters.size())
            return null;
        else{
            return shieldBoosters.remove(s);
        }
    }
        
    @Override
    public String toString(){
        String res="Esta instancia de la clase Hangar contiene las siguientes armas:\n "+weapons.toString()+
                "  \ny los siguientes potenciadores de escudo:\n"+shieldBoosters.toString()+"\ny puede almacenar "+maxElements+
                " elementos.\n";
        return res;
    }
    
    public Hangar copy(){
        return new Hangar(this);
    }
    
}

File: C2/PDOO/Practicas/DeepSpaceRuby/lib/Hangar.rb

#@author Pedro Pablo Ruiz Huertas y Juan Antonio Villegas Recio

#encoding:utf-8

require_relative "WeaponType"
require_relative "Weapon"
require_relative "ShieldBooster"
require_relative "HangarToUI"

module Deepspace
  class Hangar
    attr_reader :maxElements, :shieldBoosters, :weapons
  
    public
  
    def initialize(capacity)
        @maxElements=capacity
        @shieldBoosters=Array.new
        @weapons=Array.new
    end
    
    def self.newCopy(h)
      hangarAux=new(h.maxElements)
      h.shieldBoosters.each { |potenciador|
        hangarAux.addShieldBooster(potenciador)
      }
      h.weapons.each{ |arma|
        hangarAux.addWeapon(arma)
      }
      return hangarAux
    end
    
    def getUIversion
      return HangarToUI.new(self)
    end
    
    private
    def spaceAvailable
      return (@shieldBoosters.length+@weapons.length)<@maxElements
    end
    
    public
    def addWeapon(w)
      @weapons.push(w) if spaceAvailable
    end

    def addShieldBooster(s)
      @shieldBoosters.push(s) if spaceAvailable
    end
    
    def removeWeapon(w)
      return weapons.delete_at(w)
    end
    
    def removeShieldBooster(s)
      return shieldBoosters.delete_at(s)
    end
    
    def to_s
      cad= "Armas: \n"
      weapons.each{ |arma|
        cad+=arma.to_s
      }
      cad+="\nPotenciadores de escudo:\n"
      shieldBoosters.each{ |potenciador|
        cad+=potenciador.to_s
      }
      cad+="\nPuede almacenar #{@maxElements} elementos."
      
      return cad
    end
    
  end
end
